File: python/app/modules/census_api.py
"""
this module is for census api access
"""

import logging  # type: ignore
from typing import TypedDict  # type: ignore

import requests  # type: ignore
import backoff  # type: ignore


@backoff.on_exception(backoff.expo,
                      (requests.exceptions.Timeout,
                       requests.exceptions.ConnectionError))
def census_api(url: str) -> dict:
    """
    this is a wrapper function to the census api.
    :param url: The current api endpoint for census data lookup
    :return: The return is the body and status code of the response
    """
    logging.info('accessing census api')
    try:
        result = requests.get(url)
    except requests.exceptions.RequestException as error:
        logging.error("Error in API call : %s", error)

    return {"json": result.json()["results"][0], "status_code": result.status_code}


class Row(TypedDict):
    """
    This data class is used for type checking by mypy and at runtime.
    """
    latitude: int
    longitude: int
    block_id: int
    state_fips: int
    state_code: str
    block_pop: int


def look_up_row(row: Row):
    """
        lookup_up_row is used to iterate on a row of data and lookup census information.
    :param row:
    :return:
    """
    logging.info('looking census data for a row')
    latitude: float = row["latitude"]
    longitude: float = row["longitude"]
    response: dict = census_api("https://geo.fcc.gov/api/census/area?lat=" +
                                str(latitude) +
                                "0&lon=" +
                                str(longitude) +
                                "&format=json")
    census_information: dict = response["json"]
    row["block_id"] = census_information["block_fips"]
    row["state_fips"] = census_information["state_fips"]
    row["state_code"] = census_information["state_code"]
    row["block_pop"] = census_information["block_pop_2015"]
    return row

File: go/data_ingestion/census_api.go
//This module is used for interacting with the census rest api

package data_ingestion

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"time"
)

// used for adjusting the timeout
var httpClient = &http.Client{Timeout: 10 * time.Second}

// Inner is the second layer of the json document
type Inner struct {
	blockId int `json:"block_id"`
	stateCode string `json:"state_code"`
	stateFips int `json:"state_fips"`
	blockPop int `json:"block_pop_2015"`
}

//Outer is the first layer of the json document
type Outer struct {
	Results []Inner `json:"results"`
}


// census_api is used for interacting indirectly with he census api
// Params:
//       latitude: number to be added to y
//       longitude: number to be added to x
//return:
//       Jason return document
//       rest http response code
//       the error
func census_api( latitude,longitude float64) (Outer,int,error){
	url := "https://geo.fcc.gov/api/census/area?lat=" + FloatToString(latitude) + "0&lon=" + FloatToString(longitude) + "&format=json"
	response, err := httpClient.Get(url)
	if err != nil {
		fmt.Print(err.Error())
		os.Exit(1)
	}
	defer response.Body.Close()
	body, err := ioutil.ReadAll(response.Body)
	if err != nil {
		fmt.Print(err.Error())
		os.Exit(1)
	}
	var census Outer
	json.Unmarshal(body, &census)
	fmt.Printf("%+v\n", census)
	return census,response.StatusCode,err
}

